### PR TITLE
Fixing Python linking error

### LINKED ISSUE #147: Errors when follow README for 0.1.0 release
Env: MacOS

I followed README to test v0.1.0 release, and have following errors, please take a look @haixuanTao 

```
(base) ~/GitHub/dora-rs/bin/test1 $ dora start dataflow.yml
dyld[35552]: Library not loaded: /Library/Frameworks/Python.framework/Versions/3.11/Python
  Referenced from: <49D90856-1186-3169-BACA-0F128A85669B> /Users/yonghe/GitHub/dora-rs/bin/dora-runtime
  Reason: tried: '/Library/Frameworks/Python.framework/Versions/3.11/Python' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Library/Frameworks/Python.framework/Versions/3.11/Python' (no such file), '/Library/Frameworks/Python.framework/Versions/3.11/Python' (no such file), '/System/Library/Frameworks/Python.framework/Versions/3.11/Python' (no such file, not in dyld cache)
dyld[35551]: Library not loaded: /Library/Frameworks/Python.framework/Versions/3.11/Python
  Referenced from: <49D90856-1186-3169-BACA-0F128A85669B> /Users/yonghe/GitHub/dora-rs/bin/dora-runtime
  Reason: tried: '/Library/Frameworks/Python.framework/Versions/3.11/Python' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Library/Frameworks/Python.framework/Versions/3.11/Python' (no such file), '/Library/Frameworks/Python.framework/Versions/3.11/Python' (no such file), '/System/Library/Frameworks/Python.framework/Versions/3.11/Python' (no such file, not in dyld cache)
Traceback (most recent call last):
  File "/Users/yonghe/GitHub/dora-rs/bin/test1/./node_1/node_1.py", line 4, in <module>
    from dora import Node
ModuleNotFoundError: No module named 'dora'
  2022-11-16T17:37:32.302484Z ERROR dora_coordinator: error occured in dataflow `48b40fe9-3a44-4a63-b952-122583f5fb4c`

Caused by:
   0: failed to run dataflow at /Users/yonghe/GitHub/dora-rs/bin/test1/dataflow.yml
   1: custom node failed
   2: runtime node op_1 failed (unknown exit code)

Location:
    binaries/coordinator/src/run/runtime.rs:55:17
    at binaries/coordinator/src/lib.rs:90

48b40fe9-3a44-4a63-b952-122583f5fb4c
(base) ~/GitHub/dora-rs/bin/test1 $ ls ../
dataflow.yml                 dora-coordinator             dora-v0.1.0-x86_64-macOS.zip test1
dora                         dora-runtime                 iceoryx
(base) ~/GitHub/dora-rs/bin/test1 $ python --version
Python 3.9.7
(base) ~/GitHub/dora-rs/bin/test1 $ pip --version
pip 21.2.4 from /usr/local/lib/python3.9/site-packages/pip (python 3.9)
(base) ~/GitHub/dora-rs/bin/test1 $ 
```

### PR DESCRIPTION
**Pull request by haixuanTao**
This commit is an initial draft at fixing #147. The error is due to the fact that pyo3 has linked the libpython from the compilation and not trying to use libpython that is available in `LD_LIBRARY_PATH`.

The current only solution to disable the linking is to use the `pyo3/extension-module` flag. This requires to make the python `runtime-node` packaged in a python library. The python `runtime-node` should also be fully compatible with the other operators in case we want hybrid runtime.

The issue that i'm facing is that. Due to the packaging, I have to deal with the `GIL` that is present from the start of `dora-runtime` node. This however makes the process single threaded wich is impossible.

So, I have to disable it, but when I do, I have a race condition:
```bash
Exception ignored in: <module 'threading' from '/usr/lib/python3.8/threading.py'>
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 1373, in _shutdown
    assert tlock.locked()
AssertionError:
```
The issue is the same as https://github.com/PyO3/pyo3/issues/1274

To fix this issue, I'm going to look again at the different step to make this work.

But this is my current investigation.

Fixes #147

### CODE REVIEW

File: examples/python-dataflow/dataflow_without_webcam.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 communication:
   zenoh:
-    prefix: /example-python-no-webcam-dataflow
+    prefix: /example-python-dataflow
```
Comment (phil-opp) [inline examples/python-dataflow/dataflow_without_webcam.yml:3]: Was this change intentional?

File: binaries/runtime/Cargo.toml [hunk 2]
```diff
@@ -32,7 +32,7 @@ tokio-stream = "0.1.8"
 zenoh = { git = "https://github.com/eclipse-zenoh/zenoh.git", rev = "79a136e4fd90b11ff5d775ced981af53c4f1071b" }
 zenoh-config = { git = "https://github.com/eclipse-zenoh/zenoh.git", rev = "79a136e4fd90b11ff5d775ced981af53c4f1071b" }
 fern = "0.6.1"
-pyo3 = { version = "0.16.5", features = ["auto-initialize", "eyre", "abi3"] }
+pyo3 = { version = "0.16", features = ["eyre", "abi3-py37"] }
```
Comment (phil-opp) [inline binaries/runtime/Cargo.toml:35]: Doesn't this tie us to Python 3.7? What happens if the target system has e.g. 3.11 installed?

File: binaries/runtime/src/operator/python.rs [hunk 3]
```diff
@@ -170,24 +169,22 @@ pub fn spawn(
         Result::<_, eyre::Report>::Ok(reason)
     };
 
-    thread::spawn(move || {
```
Comment (phil-opp) [inline binaries/runtime/src/operator/python.rs:173]: With this change, the `spawn` function blocks until the operator finishes, doesn't it? So spawning multiple operators will executate them sequentially, and also block the runtime.

File: binaries/runtime/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    for operator_config in &operators {
+        let events_tx = operator_events_tx.get(&operator_config.id).unwrap();
+        spawn_operator(
+            &node_id,
+            operator_config.clone(),
+            events_tx.clone(),
+            communication.as_mut(),
+        )
+        .wrap_err_with(|| format!("failed to init operator {}", operator_config.id))?;
```
Comment (phil-opp) [inline binaries/runtime/src/lib.rs:93]: Because of  the changes to `operator/python.rs`, this `spawn_operator` call will block until the operator is finished for Python operators. So operators that should be launched after the first Python operator would be delayed until that operator is finished. Or am I missing something?

File: binaries/runtime/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap()
+            .block_on(run(
+                node_id_clone,
+                operator_events,
+                operator_stop_publishers,
+            ))
+            .unwrap();
```
Comment (phil-opp) [inline binaries/runtime/src/lib.rs:82]: I don't like the `unwrap` calls here. Errors in the `run` function are turned into panics this way, which are much more difficult to deal with. Also, we don't know about any errors until we call `stop_thread.join()` in the end, which can lead to deadlocks when an operator waits for a `stopped` message that won't be sent anymore (as the `run` loop is no longer running).

File: binaries/runtime/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    while let Some(event) = events.next().await {
+        match event {
+            Event::Operator { id, event } => {
+                match event {
+                    OperatorEvent::Error(err) => {
+                        bail!(err.wrap_err(format!("operator {id} failed")))
+                    }
+                    OperatorEvent::Panic(payload) => std::panic::resume_unwind(payload),
+                    OperatorEvent::Finished { reason } => {
+                        if let Some(stop_publisher) = operator_stop_publishers.remove(&id) {
```
Comment (phil-opp) [inline binaries/runtime/src/lib.rs:130]: I think you accidentally removed support for the `ExplicitStopAll` handling introduced in #134 here. See https://github.com/dora-rs/dora/pull/134/files#diff-6a4df06248f886d7c9a17ad464fbd9aec8e6f6269b51946a5312f5e8c51274eeR113-R120.

File: binaries/runtime/src/operator/python.rs (hunk 3 above)
Comment (haixuanTao) [reply binaries/runtime/src/operator/python.rs:173]: Yes. Basically, to make it work. I have to not spawn the python operator. And instead run the python operator in the main thread. I had to make the main thread that used to contain the "stopping thread" spawn in another thread. This ultimately makes the runtime only usable with one python operator. 

This only applies for python operator. You can spawn other rust/c operator before python for hybrid runtime

File: binaries/runtime/src/lib.rs (hunk 6 above)
Comment (haixuanTao) [reply binaries/runtime/src/lib.rs:130]: As I'm spawning the "stopping thread" in another thread than the main thread, I have to borrow the communication layer for static. And I'm not sure how to get around it.

File: binaries/runtime/src/lib.rs (hunk 5 above)
Comment (haixuanTao) [reply binaries/runtime/src/lib.rs:82]: Ok, I'll remove them

File: binaries/runtime/src/lib.rs (hunk 4 above)
Comment (haixuanTao) [reply binaries/runtime/src/lib.rs:93]: Yes. Maybe if I join the python operator at the end. It could be possible to have multiple python operator on the same runtime.

File: binaries/runtime/Cargo.toml (hunk 2 above)
Comment (haixuanTao) [reply binaries/runtime/Cargo.toml:35]: This means, that the minimum version should be 3.7, but will work with 3.11 :)

File: binaries/runtime/src/lib.rs (hunk 6 above)
Comment (haixuanTao) [reply binaries/runtime/src/lib.rs]: Oh actually we can, by replacing the `manual_stop_publisher` from a `FnOnce` closure to a `publisher`. This makes `communication` only borrowed for the time needed to create a `publisher`.

File: binaries/runtime/src/lib.rs (hunk 5 above)
Comment (haixuanTao) [reply binaries/runtime/src/lib.rs]: The race condition you're mentioning seems to be unavoidable with Python and the `GIL`. Avoiding it would require spawning the python operator in another thread which is impossible.

Is it ok to leave it as is for now?

File: binaries/runtime/Cargo.toml (hunk 2 above)
Comment (phil-opp) [reply binaries/runtime/Cargo.toml:35]: Ah, I skipped over the word "minimum" in https://pyo3.rs/latest/features.html#the-abi3-pyxy-features . Thanks for clarifying!

File: binaries/runtime/src/lib.rs (hunk 5 above)
Comment (phil-opp) [reply binaries/runtime/src/lib.rs]: I guess it's ok for now, given that we're mostly working with runtimes that run a single Python operator.

**Review by phil-opp — APPROVED** [review]
Thanks for the updates!

As I understand it, the runtime now only works with a single Python operator. If a Python operator is combined with other operators, Python or Rust, it might not spawn the additional operators (because it waits until the Python operator finishes first). Is that correct? If so, I think we should forbid spawning Python operators together with other operators for now and throw an error in that case.

Otherwise, this PR seems good to go.

**Comment by haixuanTao** [comment]
> Thanks for the updates!
> 
> As I understand it, the runtime now only works with a single Python operator. If a Python operator is combined with other operators, Python or Rust, it might not spawn the additional operators (because it waits until the Python operator finishes first). Is that correct? If so, I think we should forbid spawning Python operators together with other operators for now and throw an error in that case.
> 
> Otherwise, this PR seems good to go.

Yes, we could engineer a way to make this error-prone but they might be some edge cases and deadlock that could happen so let's forbid spawning python operators with other operators.
